src/v/archival/ntp_archiver_service.cc

```diff
@@ -871,6 +871,24 @@ ss::future<cloud_storage::upload_result> ntp_archiver::upload_manifest(
     co_return result;
 }
 
+int64_t ntp_archiver::records_pending_upload() const {
+    const auto hwm = _parent.high_watermark();
+
+    if (hwm == model::offset{}) {
+        return 0;
+    }
+
+    const auto ot_state = _parent.get_offset_translator_state();
+    const auto kafka_hwm = ot_state->from_log_offset(hwm);
+
+    auto last_uploaded_kafka_offset = manifest().get_last_kafka_offset();
+    if (!last_uploaded_kafka_offset) {
+        return kafka_hwm();
```

Comment:
Do we need to subtract 1? Kafka's HWM is one past the end. Or perhaps should we be translating `hwm - 1`?

Reply:
I don't think we need to subtract. If HWM=5 and no offset was yet uploaded, then there's 5 records to upload: [0 ... 4]